README.md
# LWNamePopularityVisualizer

[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
![GitHub release (latest by date)](https://img.shields.io/github/v/release/leeway64/LWNamePopularityVisualizer)

LWNamePopularityVisualizer visualizes the popularity of a certain name for each decade.


## Usage

```
$ python3 -m venv .venv
$ source .venv/bin/activate
$ pip3 install -r requirements.txt
$ python3 src/LWNamePopularityVisualizer.py include/names.txt


This program allows you to search through the
data from the Social Security Administration
to see how popular a particular name has been
since 1890.
Name: jon
Gender (M or F): m
Jon    M  0  0 0 979 672 123 138 	97  68  127 216 354 576

JON m Scandinavian, Basque Scandinavian and Basque form of JOHN; m English Short form of ATHAN
```

Which will generate the following plot:

![Jon name popularity plot](doc/jon_popularity_1890.png)

The y-axis values represents how popular a given name was for each decade. For example, a
popularity of 1 (most popular) would mean that that name was the 1st most popular name for that decade.
A popularity of 1000 (least popular) would mean that that name was the 1000th most popular name for that
decade.

The default name popularity file is [`names.txt`](include/names.txt) with a start year of 1890. If
the name popularity file is set to [`names2.txt`](include/names2.txt), then the start year will be
1863.

If names2.txt had been provided, then:

```
$ python3 src/LWNamePopularityVisualizer.py include/names2.txt


This program allows you to search through the
data from the Social Security Administration
to see how popular a particular name has been
since 1863.
Name: myra
Gender (M or F): f
Myra   f 209  254 309 289 300  225  236 261

MYRA f English Created by the 17th-century poet Fulke Greville.
```
![Myra name popularity plot](doc/myra_popularity_1863.png)


## Unit Tests

To run unit tests, run:
```
pytest
```


## Third-Party Software

- [Matplotlib](https://matplotlib.org/): Plotting and visualization library for Python.
- [pytest](https://docs.pytest.org/en/7.0.x/) (MIT license): Python unit testing framework.


## References and Acknowledgements

The name popularity information in [`names.txt`](include/names.txt) and [`names2.txt`](include/names2.txt)
is based on [data from the Social Security Administration](https://www.ssa.gov/OACT/babynames/).

The specifications for this project, along with `names.txt` and `names2.txt`, was provided on the
[UW Seattle CSE 142 summer 2017 website](https://courses.cs.washington.edu/courses/cse142/17su/homework.shtml).
The original assignment was meant to be written in Java, but I completed it using Python.

This project also includes a [Python class](src/HashSet.py) that provides a basic implementation
for a HashSet.

I would like to give credit to UW Seattle's autumn 2017 CSE 143 course for providing the basis for
this code. The original HashSet (written in Java) can be found on
[this page](https://courses.cs.washington.edu/courses/cse143/17au/lectures/HashSet.java);
I based my Python implementation on the Java implementation.


CHANGELOG.md
# Changelog

All changes between versions for this project will be documented in this file.

This changelog format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and
this project follows [semantic versioning](https://semver.org/).

## Unreleased

- None

## [1.1.0] - 2023-8-26
### Added
- HashSet
- Unit tests for HashSet

### Changed
- LWNamePopularityVisualizer now searches for the names file from the current working directory

## [1.0.0] - 2022-5-1

- First release



src/test_suite.py
from HashSet import HashSet


class TestSuite:
    def test_add_remove_size(self):
        set1 = HashSet(5000)
        assert set1.size() == 0

        set1.add(101)
        assert set1.size() == 1
        set1.remove(101)
        assert set1.empty() == True
        assert set1.size() == 0

        set1.add(1,2,3,4)
        assert set1.size() == 4
        set1.remove(2)
        assert set1.size() == 3

    def test_contains_equals(self):
        set1 = HashSet(120)
        set2 = HashSet(240)

        assert not set1.contains(456)
        assert not set2.contains(7894231)
        assert set1 == set2

        set1.add(1)
        set1.add(2)
        set1.add(4)
        set1.add(8)
        set1.add(16)
        assert set1.contains(16)

        set2.add(1)
        set2.add(2)
        set2.add(4)
        set2.add(8)
        set2.add(16)
        assert set1 == set2
        set2.add(32)
        assert set1 != set2
        assert set2.contains(32)
        assert set2.contains(64) == False
    
    def test_strings(self):
        set1 = HashSet(800)
        
        set1.add("a")
        set1.add("b")
        set1.add("c")
        set1.add("a")
        assert set1.size() == 3  # No duplicates are allowed
        assert set1.contains("c")

        set1.remove("b")
        assert set1.contains("b") != True


test/test_suite.py
from Basic_HashSet_Implementation.HashSet import HashSet


class TestSuite:
    def test_add_remove_size(self):
        set1 = HashSet(5000)
        assert set1.size() == 0

        set1.add(101)
        assert set1.size() == 1
        set1.remove(101)
        assert set1.empty() == True
        assert set1.size() == 0

        set1.add(1,2,3,4)
        assert set1.size() == 4
        set1.remove(2)
        assert set1.size() == 3

    def test_contains_equals(self):
        set1 = HashSet(120)
        set2 = HashSet(240)

        assert not set1.contains(456)
        assert not set2.contains(7894231)
        assert set1 == set2

        set1.add(1)
        set1.add(2)
        set1.add(4)
        set1.add(8)
        set1.add(16)
        assert set1.contains(16)

        set2.add(1)
        set2.add(2)
        set2.add(4)
        set2.add(8)
        set2.add(16)
        assert set1 == set2
        set2.add(32)
        assert set1 != set2
        assert set2.contains(32)
        assert set2.contains(64) == False
    
    def test_strings(self):
        set1 = HashSet(800)
        
        set1.add("a")
        set1.add("b")
        set1.add("c")
        set1.add("a")
        assert set1.size() == 3  # No duplicates are allowed
        assert set1.contains("c")

        set1.remove("b")
        assert set1.contains("b") != True


src/HashSet.py
# An implementation of a hash set

class HashSet:
    def __init__(self, capacity):
        self.__data = []
        self.__size = 0  # Number of elements in the HashSet
        self.__hash_table_length = int(capacity / 0.75)  # Load factor: 0.75
        self.__initialize_data(self.__hash_table_length)

    # Add multiple values to the HashSet. Values can be as many values as needed.
    def add(self, *values):
        for value in values:
            if not self.contains(value):  # Does not allow any duplicates
                i = self.__index_of(value)
                self.__data[i] = self.__HashNode(value, self.__data[i])
                self.__size += 1

    # Remove value from the HashSet, assuming value is in the HashSet
    def remove(self, value):
        if self.contains(value):
            i = self.__index_of(value)
            if self.__data[i].value == value:
                self.__data[i] = self.__data[i].next
            else:
                current = self.__data[i]
                while current.next.value != value:
                    current = current.next
                current.next = current.next.next
            self.__size -= 1

    # Checks if the HashSet contains the value
    # Returns True if it does contain value, False if not
    def contains(self, value):
        i = self.__index_of(value)
        current = self.__data[i]
        while current is not None:
            if current.value == value:
                return True
            current = current.next
        return False

    # Returns the number of elements in the HashSet
    def size(self):
        return self.__size

    def empty(self):
        return self.__size == 0

    # Returns the location (the index) of the value in the HashSet
    def __index_of(self, value):
        return int(abs(hash(value) % self.__hash_table_length))  # Hash code is the same for ints

    # Creates a list that is table_length long. Each list element is a HashNode, since this is a
    # list of linked lists
    def __initialize_data(self, table_length):
        for i in range(table_length):
            self.__data.append(self.__HashNode())

    # Returns a string version of the HashSet in an easily readable format
    def __str__(self):
        result = "{\n"
        for i in range(self.__hash_table_length):
            if self.__data[i].value is not None:
                current = self.__data[i]
                while current.value is not None:
                    result += str(i) + ': ' + str(current.value) + '\n'
                    current = current.next
        result += "}"
        return result

    # Returns an iterator composed of the values in this HashSet
    def __iter__(self):
        result = []
        for i in range(self.__hash_table_length):
            if self.__data[i].value is not None:
                current = self.__data[i]
                while current.value is not None:
                    result.append(current.value)
                    current = current.next
        
        return iter(result)

    # Returns true if self and other have the same elements
    def __eq__(self, other):
        return list(iter(self)) == list(iter(other))

    # 1 node of a linked list
    class __HashNode:
        def __init__(self, value=None, next=None):
            self.value = value
            self.next = next


src/LWNamePopularityVisualizer.py
# This programs prompts the user for a name and gender. If information on the name (with the respective gender)
# exists, then the program will print the popularity and name meaning data. It will also draw a plot
# representing the name's popularity.

from matplotlib import pyplot as plt
import os
import sys

from HashSet import HashSet


BAR_WIDTH = 3


# Prints an introductory message.
def introduction(START_YEAR):
    print("This program allows you to search through the")
    print("data from the Social Security Administration")
    print("to see how popular a particular name has been")
    print("since " + str(START_YEAR) + ".")


# Searches the names or meanings file for the specific name and returns the entire line that contains the
# name.
# Inputs -
# name (string): the name to search for
# gender (string): the gender of the name to search for
# text_file (string): the text file to search through
# Returns -
# The program returns the entire line that contains the name that it's searching for
def search_file(name, gender, text_file):
    with open(text_file) as file:
        line = file.readline()
        split_line = line.lower().split()
        while line:
            if split_line[0] == name.lower() and split_line[1] == gender.lower():
                return line
            line = file.readline()
            split_line = line.lower().split()

        return "\"" + name + "\" not found"


# Creates a bar chart based on the popularity of each name.
# Inputs -
# name_popularity (list): a list of numbers representing the popularity of each name for every decade
# name_meaning (string): the meaning of the name
# No returns
def draw_bar_chart(name_popularity, name_meaning, START_YEAR):
    popularity_data = name_popularity.split()
    popularity_data = popularity_data[2:]
    popularity_data = [int(z) for z in popularity_data]

    # Popularity is mapped to the y-axis using the equation y_data = -popularity_data + 1001, assuming
    # popularity_data is not 0. If popularity_data is 0, then y_data would also be 0.
    # Basically, a popularity of 1 should map to a height of 1000, while a popularity of 1000 should map to
    # a height of 1.
    y_data = []
    for i in range(len(popularity_data)):
        if popularity_data[i] != 0:
            y_data.append((-1*popularity_data[i])+1001)
        else:
            y_data.append(0)

    year_data = [x*10 for x in range(len(popularity_data))]
    year_data = [y+START_YEAR for y in year_data]

    plt.style.use('fivethirtyeight')
    plt.figure(name_meaning)
    plt.bar(year_data, y_data, width=BAR_WIDTH)
    plt.title('Popularity of name vs. years')

    plt.xlabel('Years (decades)')
    plt.xticks(ticks=year_data, labels=year_data)

    plt.ylabel('Popularity of name (nth most popular)')
    plt.yticks(ticks=[])

    for i in range(len(year_data)):
        plt.text(year_data[i], y_data[i], str(popularity_data[i]))

    plt.grid(True)
    plt.show()


def main():    
    if len(sys.argv) != 2:
        raise RuntimeError("Incorrect number of arguments! Provide the name of the file that contains the names.")
    
    
    name_file = f'{sys.argv[1]}'
    START_YEAR = 1890
    if "names2.txt" in sys.argv[1]:
        START_YEAR = 1863
        
    meanings_file = f"include/meanings.txt"

    hash_set = HashSet(1945)
    hash_set.add(1945)
    if hash_set.contains(1945):
        introduction(START_YEAR)
    
    name = input("Name: ")
    gender = input("Gender (M or F): ")

    name_popularity = search_file(name, gender, name_file)
    print(name_popularity)

    if name_popularity != "\"" + name + "\" not found":
        name_meaning = search_file(name, gender, meanings_file)
        print(name_meaning)
        draw_bar_chart(name_popularity, name_meaning, START_YEAR)


if __name__ == '__main__':
    main()



main.py
# This programs prompts the user for a name and gender. If information on the name (with the respective gender)
# exists, then the program will print the popularity and name meaning data. It will also draw a plot
# representing the name's popularity.

from matplotlib import pyplot as plt


START_YEAR = 1890
# START_YEAR = 1863
BAR_WIDTH = 3


# Prints an introductory message.
def introduction():
    print("This program allows you to search through the")
    print("data from the Social Security Administration")
    print("to see how popular a particular name has been")
    print("since " + str(START_YEAR) + ".")


# Searches the names or meanings file for the specific name and returns the entire line that contains the
# name.
# Inputs -
# name (string): the name to search for
# gender (string): the gender of the name to search for
# text_file (string): the text file to search through
# Returns -
# The program returns the entire line that contains the name that it's searching for
def search_file(name, gender, text_file):
    with open(text_file) as file:
        line = file.readline()
        split_line = line.lower().split()
        while line:
            if split_line[0] == name.lower() and split_line[1] == gender.lower():
                return line
            line = file.readline()
            split_line = line.lower().split()

        return "\"" + name + "\" not found"


# Creates a bar chart based on the popularity of each name.
# Inputs -
# name_popularity (list): a list of numbers representing the popularity of each name for every decade
# name_meaning (string): the meaning of the name
# No returns
def draw_bar_chart(name_popularity, name_meaning):
    popularity_data = name_popularity.split()
    popularity_data = popularity_data[2:]
    popularity_data = [int(z) for z in popularity_data]

    # Popularity is mapped to the y-axis using the equation y_data = -popularity_data + 1001, assuming
    # popularity_data is not 0. If popularity_data is 0, then y_data would also be 0.
    # Basically, a popularity of 1 should map to a height of 1000, while a popularity of 1000 should map to
    # a height of 1.
    y_data = []
    for i in range(0, len(popularity_data)):
        if popularity_data[i] != 0:
            y_data.append((-1*popularity_data[i])+1001)
        else:
            y_data.append(0)

    year_data = [x*10 for x in range(0, len(popularity_data))]
    year_data = [y+START_YEAR for y in year_data]

    plt.style.use('fivethirtyeight')
    plt.figure(name_meaning)
    plt.bar(year_data, y_data, width=BAR_WIDTH)
    plt.title('Popularity of name vs. years')

    plt.xlabel('Years (decades)')
    plt.xticks(ticks=year_data, labels=year_data)

    plt.ylabel('Popularity of name (nth most popular)')
    plt.yticks(ticks=[])

    for i in range(0, len(year_data)):
        plt.text(year_data[i], y_data[i], str(popularity_data[i]))

    plt.grid(True)
    plt.show()


def main():
    name_file = 'names.txt'  # Start date for first name file is 1890
    # name_file = "names2.txt"  # Start date for second name file is 1863
    meanings_file = "meanings.txt"


    introduction()
    name = input("Name: ")
    gender = input("Gender (M or F): ")

    name_popularity = search_file(name, gender, name_file)
    print(name_popularity)

    if name_popularity != "\"" + name + "\" not found":
        name_meaning = search_file(name, gender, meanings_file)
        print(name_meaning)
        draw_bar_chart(name_popularity, name_meaning)


if __name__ == '__main__':
    main()
